src/snowflake/snowpark/_internal/analyzer/select_statement.py

```diff
@@ -499,46 +499,25 @@ def select(self, cols: List[Expression]) -> "SelectStatement":
         elif self.flatten_disabled or self.has_clause_using_columns:
             can_be_flattened = False
         else:
-            can_be_flattened = True
-            subquery_column_states = self.column_states
+            can_be_flattened = can_select_statement_be_flattened(
+                self.column_states, new_column_states
+            )
+
+        if can_be_flattened:
+            new = copy(self)
+            final_projection = []
+
             for col, state in new_column_states.items():
-                dependent_columns = state.dependent_columns
-                if dependent_columns == COLUMN_DEPENDENCY_DOLLAR:
-                    can_be_flattened = False
-                    break
-                subquery_state = subquery_column_states.get(col)
                 if state.change_state in (
                     ColumnChangeState.CHANGED_EXP,
                     ColumnChangeState.NEW,
                 ):
-                    can_be_flattened = can_projection_dependent_columns_be_flattened(
-                        dependent_columns, subquery_column_states
-                    )
-                    if not can_be_flattened:
-                        break
                     final_projection.append(copy(state.expression))
                 elif state.change_state == ColumnChangeState.UNCHANGED_EXP:
-                    # query may change sequence of columns. If subquery has same-level reference, flattened sql may not work.
-                    if (
-                        col not in subquery_column_states
-                        or subquery_column_states[col].depend_on_same_level
-                    ):
-                        can_be_flattened = False
-                        break
                     final_projection.append(
-                        copy(subquery_column_states[col].expression)
+                        copy(self.column_states[col].expression)
```

Comment:
This is a logic change

Reply:
I deleted line 503 `subquery_column_states = self.column_states` , so these should be equivalent?